### PR TITLE
AWS: add skip name validation config for glue namespace

### PR DESCRIPTION
**Pull request by xingfanx**
Add a configuration `awsProperties.glueCatalogSkipNameValidation()` which allows characters like `-` in glue catalog namespaces

Tested with unittests.
```
> Task :iceberg-aws:test
BUILD SUCCESSFUL in 9s
13 actionable tasks: 1 executed, 12 up-to-date
4:07:42 PM: Execution finished ':iceberg-aws:test --tests "org.apache.iceberg.aws.glue.TestIcebergToGlueConverter"'.
```

### CODE REVIEW

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 1]
```diff
@@ -103,6 +103,12 @@ public class AwsProperties implements Serializable {
   public static final String GLUE_CATALOG_SKIP_ARCHIVE = "glue.skip-archive";
   public static final boolean GLUE_CATALOG_SKIP_ARCHIVE_DEFAULT = false;
 
+  /**
+   * If Glue should skip name validations
+   */
+  public static final String GLUE_CATALOG_SKIP_NAME_VALIDATION = "glue.skip-archive";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:109]: this config name is not right

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 2]
```diff
@@ -464,8 +467,9 @@ public boolean setProperties(Namespace namespace, Map<String, String> properties
     newProperties.putAll(properties);
     glue.updateDatabase(UpdateDatabaseRequest.builder()
         .catalogId(awsProperties.glueCatalogId())
-        .name(IcebergToGlueConverter.toDatabaseName(namespace))
-        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, newProperties))
+        .name(IcebergToGlueConverter.toDatabaseName(namespace, awsProperties.glueCatalogSkipNameValidation()))
+        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, newProperties,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:471]: nit: prefer all parameters on the same line when breaking a newline,

```
.databaseInput(IcebergToGlueConverter.toDatabaseInput(
  namespace, newProperties,awsProperties.glueCatalogSkipNameValidation()))
```

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 3]
```diff
@@ -481,8 +485,9 @@ public boolean removeProperties(Namespace namespace, Set<String> properties) thr
 
     glue.updateDatabase(UpdateDatabaseRequest.builder()
         .catalogId(awsProperties.glueCatalogId())
-        .name(IcebergToGlueConverter.toDatabaseName(namespace))
-        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, metadata))
+        .name(IcebergToGlueConverter.toDatabaseName(namespace, awsProperties.glueCatalogSkipNameValidation()))
+        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, metadata,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:489]: same comment for breaking newline

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java [hunk 4]
```diff
@@ -74,7 +74,8 @@ class GlueTableOperations extends BaseMetastoreTableOperations {
                       FileIO fileIO, TableIdentifier tableIdentifier) {
     this.glue = glue;
     this.awsProperties = awsProperties;
-    this.databaseName = IcebergToGlueConverter.getDatabaseName(tableIdentifier);
+    this.databaseName = IcebergToGlueConverter.getDatabaseName(tableIdentifier,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:77]: same comment for breaking newline

File: aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java [hunk 5]
```diff
@@ -105,31 +105,41 @@ static void validateNamespace(Namespace namespace) {
...
-   * @param metadata metadata map
+   *
+   * @param namespace                Iceberg namespace
+   * @param metadata                 metadata map
+   * @param shouldSkipNameValidation should skip name validation
    * @return Glue DatabaseInput
    */
-  static DatabaseInput toDatabaseInput(Namespace namespace, Map<String, String> metadata) {
-    DatabaseInput.Builder builder = DatabaseInput.builder().name(toDatabaseName(namespace));
+  static DatabaseInput toDatabaseInput(Namespace namespace, Map<String, String> metadata,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:139]: same comment for breaking newline

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 6]
```diff
@@ -450,7 +453,7 @@ public boolean dropNamespace(Namespace namespace) throws NamespaceNotEmptyExcept
 
     glue.deleteDatabase(DeleteDatabaseRequest.builder()
         .catalogId(awsProperties.glueCatalogId())
-        .name(IcebergToGlueConverter.toDatabaseName(namespace))
+        .name(IcebergToGlueConverter.toDatabaseName(namespace, awsProperties.glueCatalogSkipNameValidation()))
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:456]: I think skipNameValidation should suffice (I don't think we need the glueCatalog part in the name)

File: aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java [hunk 7]
```diff
@@ -105,31 +105,41 @@ static void validateNamespace(Namespace namespace) {
...
    * Validate and convert Iceberg namespace to Glue database name
-   * @param namespace Iceberg namespace
+   *
+   * @param namespace                Iceberg namespace
+   * @param shouldSkipNameValidation should skip name validation
    * @return database name
    */
-  static String toDatabaseName(Namespace namespace) {
-    validateNamespace(namespace);
+  static String toDatabaseName(Namespace namespace, boolean shouldSkipNameValidation) {
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:113]: Similar to above, I think the boolean can just be skipNameValidation

File: aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java [hunk 8]
```diff
@@ -105,31 +105,41 @@ static void validateNamespace(Namespace namespace) {
...
+    if (!shouldSkipNameValidation) {
+      validateNamespace(namespace);
+    }
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/IcebergToGlueConverter.java:116]: newline after if block

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 9]
```diff
@@ -201,7 +201,7 @@ protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
       Map<String, String> tableSpecificCatalogProperties = ImmutableMap.<String, String>builder()
           .putAll(catalogProperties)
           .put(AwsProperties.LAKE_FORMATION_DB_NAME,
-              IcebergToGlueConverter.getDatabaseName(tableIdentifier))
+              IcebergToGlueConverter.getDatabaseName(tableIdentifier, awsProperties.glueCatalogSkipNameValidation()))
           .put(AwsProperties.LAKE_FORMATION_TABLE_NAME,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:205]: I think we have to also skip table name validation in the same way using the same config

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 10]
```diff
@@ -103,6 +103,12 @@ public class AwsProperties implements Serializable {
   public static final String GLUE_CATALOG_SKIP_ARCHIVE = "glue.skip-archive";
   public static final boolean GLUE_CATALOG_SKIP_ARCHIVE_DEFAULT = false;
 
+  /**
+   * If Glue should skip name validations
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:107]: add some more explanations here, it is recommended to stick to Glue best practice in https://docs.aws.amazon.com/athena/latest/ug/glue-best-practices.html to make sure operations are Hive compatible. This is only added for services that have existing conventions using non-standard characters. When database name and table name validation is skipped, there is no guarantee that downstream systems would all support the names.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java (hunk 6 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:456]: actually we have to, because we already have `glueCatalogSkipArchive()`...

Comment (amogh-jahagirdar) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:456]: Ah good point, nvm @xingfanx !

File: docs/integrations/aws.md [hunk 11]
```diff
@@ -180,6 +181,14 @@ However, if you are streaming data to Iceberg, this will easily create a lot of
 Therefore, it is recommended to turn off the archive feature in Glue by setting `glue.skip-archive` to `true`.
 For more details, please read [Glue Quotas](https://docs.aws.amazon.com/general/latest/gr/glue.html) and the [UpdateTable API](https://docs.aws.amazon.com/glue/latest/webapi/API_UpdateTable.html).
 
+#### Skip Name Validation
+
+Allow user to skip name validation for table name and namespaces.
+It is recommended to stick to Glue best practice in
+https://docs.aws.amazon.com/athena/latest/ug/glue-best-practices.html to make sure operations are Hive compatible.
+This is only added for services that have existing conventions using non-standard characters. When database name
```
Comment (jackye1995) [inline docs/integrations/aws.md:189]: not "for services", but "for users"

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 12]
```diff
@@ -103,6 +103,16 @@ public class AwsProperties implements Serializable {
...
+   * It is recommended to stick to Glue best practice in
+   * https://docs.aws.amazon.com/athena/latest/ug/glue-best-practices.html to make sure operations are Hive compatible.
+   * This is only added for services that have existing conventions using non-standard characters. When database name
+   * and table name validation is skipped, there is no guarantee that downstream systems would all support the names.
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:111]: are skipped

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 13]
```diff
@@ -103,6 +103,16 @@ public class AwsProperties implements Serializable {
   public static final String GLUE_CATALOG_SKIP_ARCHIVE = "glue.skip-archive";
   public static final boolean GLUE_CATALOG_SKIP_ARCHIVE_DEFAULT = false;
 
+  /**
+   * If Glue should skip name validations
+   * It is recommended to stick to Glue best practice in
+   * https://docs.aws.amazon.com/athena/latest/ug/glue-best-practices.html to make sure operations are Hive compatible.
+   * This is only added for services that have existing conventions using non-standard characters. When database name
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:110]: for users

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 14]
```diff
@@ -399,8 +410,8 @@ public AwsProperties() {
 
     this.glueCatalogId = null;
     this.glueCatalogSkipArchive = GLUE_CATALOG_SKIP_ARCHIVE_DEFAULT;
+    this.glueCatalogSkipNameValidation = GLUE_CATALOG_SKIP_NAME_VALIDATION_DEFAULT;
     this.glueLakeFormationEnabled = GLUE_LAKEFORMATION_ENABLED_DEFAULT;
-
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:403]: nit: should not remove empty lines if not necessary, easy to cause merge conflict for other people's PRs

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 15]
```diff
@@ -276,7 +276,8 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
       TableMetadata lastMetadata = ops.current();
       glue.deleteTable(DeleteTableRequest.builder()
           .catalogId(awsProperties.glueCatalogId())
-          .databaseName(IcebergToGlueConverter.getDatabaseName(identifier))
+          .databaseName(IcebergToGlueConverter.getDatabaseName(identifier,
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:279]: nit: newline before `identifier` instead of like this

**Review by jackye1995 — APPROVED** [review]
thanks for the fix!

**Comment by jackye1995** [comment]
@amogh-jahagirdar any other comments?

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java [hunk 16]
```diff
@@ -74,8 +74,10 @@ class GlueTableOperations extends BaseMetastoreTableOperations {
                       FileIO fileIO, TableIdentifier tableIdentifier) {
     this.glue = glue;
     this.awsProperties = awsProperties;
-    this.databaseName = IcebergToGlueConverter.getDatabaseName(tableIdentifier);
-    this.tableName = IcebergToGlueConverter.getTableName(tableIdentifier);
+    this.databaseName = IcebergToGlueConverter.getDatabaseName(
+            tableIdentifier, awsProperties.glueCatalogSkipNameValidation());
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java:78]: I think we need to check indentation here (continuing should be 4)

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 17]
```diff
@@ -359,7 +361,8 @@ public void createNamespace(Namespace namespace, Map<String, String> metadata) {
     try {
       glue.createDatabase(CreateDatabaseRequest.builder()
           .catalogId(awsProperties.glueCatalogId())
-          .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, metadata))
+          .databaseInput(IcebergToGlueConverter.toDatabaseInput(
+                  namespace, metadata, awsProperties.glueCatalogSkipNameValidation()))
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:365]: Same here, indentation

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 18]
```diff
@@ -276,7 +276,8 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
       TableMetadata lastMetadata = ops.current();
       glue.deleteTable(DeleteTableRequest.builder()
           .catalogId(awsProperties.glueCatalogId())
-          .databaseName(IcebergToGlueConverter.getDatabaseName(identifier))
+          .databaseName(IcebergToGlueConverter.getDatabaseName(
+                  identifier, awsProperties.glueCatalogSkipNameValidation()))
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:280]: indentation 4

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 19]
```diff
@@ -481,8 +486,9 @@ public boolean removeProperties(Namespace namespace, Set<String> properties) thr
 
     glue.updateDatabase(UpdateDatabaseRequest.builder()
         .catalogId(awsProperties.glueCatalogId())
-        .name(IcebergToGlueConverter.toDatabaseName(namespace))
-        .databaseInput(IcebergToGlueConverter.toDatabaseInput(namespace, metadata))
+        .name(IcebergToGlueConverter.toDatabaseName(namespace, awsProperties.glueCatalogSkipNameValidation()))
+        .databaseInput(IcebergToGlueConverter.toDatabaseInput(
+                namespace, metadata, awsProperties.glueCatalogSkipNameValidation()))
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:491]: indentation 4

**Comment by amogh-jahagirdar** [comment]
@jackye1995 just some indentation stuff, the logic and the docs look good to me. I think my comments should cover where I think the continuing indentation needs to be updated, but let me know if I missed places

**Review by amogh-jahagirdar — APPROVED** [review]

**Comment by amogh-jahagirdar** [comment]
LGTM thanks for the contribution!

**Pull request merged**